torch/distributions/lowrank_multivariate_normal.py

```diff
@@ -0,0 +1,176 @@
+    Wt_Dinv = W.transpose(-1, -2) / D.unsqueeze(-2)
+    Wt_Dinv_x = _batch_mv(Wt_Dinv, x)
+    mahalanobis_term1 = (x.pow(2) / D).sum(-1)
+    mahalanobis_term2 = _batch_mahalanobis(capacitance_tril, Wt_Dinv_x)
+    return mahalanobis_term1 - mahalanobis_term2
+
+
+class LowRankMultivariateNormal(Distribution):  # TODO: docs, tests
+    r"""
+    Creates a multivariate normal distribution with covariance matrix having a low-rank form
+    parameterized by `scale_factor` and `scale_diag`::
+        covariance_matrix = scale_factor @ scale_factor.T + scale_diag
+
+    Example:
+
+        >>> m = MultivariateNormal(torch.zeros(2), torch.tensor([1, 0]), torch.tensor([1, 1]))
+        >>> m.sample()  # normally distributed with mean=`[0,0]`, scale_factor=`[1,0]`, scale_diag=`[1,1]`
+        tensor([-0.2102, -0.5429])
+
+    Args:
+        loc (Tensor): mean of the distribution
+        scale_factor (Tensor): factor part of low-rank form of covariance matrix
+        scale_diag (Tensor): diagonal part of low-rank form of covariance matrix
+
+    Note:
+        The computation for determinant and inverse of covariance matrix is saved when
+        `scale_factor.shape[1] << scale_factor.shape[0]` thanks to "Woodbury matrix identity" and
+        "matrix determinant lemma". Thanks to these formulas, we just need to compute the
+        determinant and inverse of the small size "capacitance" matrix::
+            capacitance = I + scale_factor.T @ inv(scale_diag) @ scale_factor
+    """
+    arg_constraints = {"loc": constraints.real,
+                       "scale_factor": constraints.real,
+                       "scale_diag": constraints.positive}
+    support = constraints.real
+    has_rsample = True
+
+    def __init__(self, loc, scale_factor, scale_diag, validate_args=None):
+        event_shape = loc.shape[-1:]
+        if scale_factor.dim() < 2:
+            raise ValueError("scale_factor must be at least two-dimensional, "
+                             "with optional leading batch dimensions")
+        if scale_factor.shape[-2:-1] != event_shape:
+            raise ValueError("scale_factor must be a batch of matrices with shape {} x m"
+                             .format(event_shape[0]))
+        if scale_diag.shape[-1:] != event_shape:
+            raise ValueError("scale_diag must be a batch of vectors with shape {}".format(event_shape))
+
+        scale_batch_shape = _get_batch_shape(scale_factor, scale_diag)
+        try:
+            batch_shape = torch._C._infer_size(loc.shape[:-1], scale_batch_shape)
+        except:
+            raise ValueError("Incompatible batch shapes: loc {}, scale_factor {}, scale_diag {}"
+                             .format(loc.shape, scale_factor.shape, scale_diag.shape))
+
+        loc_shape = batch_shape + event_shape
+        self.loc = loc.expand(loc_shape)
+        self.scale_factor = scale_factor.expand(loc_shape + scale_factor.shape[-1:])
+        self.scale_diag = scale_diag.expand(loc_shape)
```

Comment:
Without these `.expand`, the `_gradcheck_log_prob` failed for LowRank implementation.